### ISSUE TITLE
Design check

### ISSUE THREAD

**Issue by ocharles** [ocharles/dhall-doc#1]
Hi @Gabriel439 (and others!),

As `dhall-to-cabal` approaches completion for its initial release, it's time to start showing why it is useful. One point of motivation was to provide a concrete specification for what it means to be a Cabal file. While the type is a good reference point, user reading documentation that is *guaranteed* to document this type is even better. `dhall-doc` is an attempt to make such documentation a reality.

My rough plan is to have `dhall-doc` essentially be a function `document : Expr -> Expr -> Documentation`. The first argument is the thing to document (anything that is a type or lambda abstraction?), and the second argument is an `Expr` that documents this type.

In my initial sketch, we have this type:

```
{ name : Text
}
```

This documentation expression

```
{ name = "The name of the thing" }
```

Which results in this document (rendered as RST here, but it produces `Pandoc`):

```rst
name : ``Text``
    The name of the thing
```

How does this sound? Do have a different approach in mind for documentation?

**Comment by ocharles** [comment]
An alternative to the "cleverness" of zipped traversals (where the names of the documentation record coincide with the names of the type being documented) would be to use a stable documentation schema:

```
< DocumentRecord : { fields : List { field-name : Text, field-documentation : Text } }
>
```

And then have `dhall-doc` know how to render this documentation out. We can the provide the ability to check that some given documentation matches type.

I think I might actually prefer this, but we need to work out what the schema of documentation actually is.

**Comment by ocharles** [comment]
One thing that is going to be crucial for `dhall-to-cabal` is the ability to reference sections. For example, `build-dependencies` is common to `library`, `executable`, `test-suite`, etc - so for the documentation of each of those types, I need to be able to say `build-dependencies : BuildDependencies`, and have `BuildDependencies` be a hyper-link to documentation.

**Comment by Gabriella439** [comment]
I was going to propose just commenting the Dhall type.  i.e.:

```haskell
{ name : Text  -- The name of the thing
...
}
```

**Comment by ocharles** [comment]
Ok, but I want something that can generate HTML documentation. Are you suggesting a Haddock like approach?

**Comment by Gabriella439** [comment]
Yeah, possibly.  I think structured comments that support markdown could be a basis for a Haddock-like tool

The main thing I'd like is if the documentation travels with the code, since that's the feature about Haddock I like the most
